pkg/build/registry/buildconfig/rest.go

```diff
@@ -75,6 +77,11 @@ func (r *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 	if errs := validation.ValidateBuildConfig(buildConfig); len(errs) > 0 {
 		return nil, errors.NewInvalid("buildConfig", buildConfig.Name, errs)
 	}
+
+	if circularDeps(buildConfig) {
```

Comment:
Running O(N) loads in maps of maps is expensive, agreed, but I suspect this can be refactored to use a cheaper structure where it won't really matter.